java/WordCountOptimized.java

```diff
@@ -217,33 +207,66 @@ public static int tokenizeAndSubmitBlock(TObjectIntCustomHashMap<byte[]> counts,
         }
         return 0;
     }
-    
-    private static void submitWord(TObjectIntCustomHashMap<byte[]> m, byte[] word) {
-        m.adjustOrPutValue(word, 1, 1);        
+
+    private static final int TABLE_SIZE = 1024 * 1024;
+    private static final int TABLE_MASK = TABLE_SIZE - 1;
+
+    static class WordMap {
+        final byte[][] keys = new byte[TABLE_SIZE][];
+        final int[] values = new int[TABLE_SIZE];
+        int size = 0;
     }
-    
+
+    private static int submitWord(byte[][] keys, int[] values, int size,
+                                   byte[] word, int off, int len) {
+        int hashCode = (int) XX_HASH.hashBytes(word, off, len);
+        int index = hashCode & TABLE_MASK;
```

Comment:
There are ~25 million unique tokens: https://github.com/juditacs/wordcount/issues/46

Support for resize may actually increase performance on some inputs -- smaller hashtables can reside largely in cache, for big performance boosts.

Tip: one fast approach to this is to use hashtable sizes that are a power of 2, and always double them.  New indices can be computed by a bitshift that way.  The hashtable should resize at 60-70% fill (you can precompute the size needed to cross that threshold with each resize operation, so it's just a greater-than check with each element added).